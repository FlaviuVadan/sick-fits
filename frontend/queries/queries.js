import gql from 'graphql-tag';

import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
    query ALL_ITEMS_QUERY(
    $skip: Int = 0
    $first: Int = ${ perPage }
    ) {
        items(
            orderBy: createdAt_DESC
            skip: $skip
            first: $first
        ) {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
    ) {
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`;

export const UPDATE_ITEM_MUTATION = gql`
    mutation UPDATE_ITEM_MUTATION(
    $id: ID!
    $title: String
    $description: String
    $price: Int
    ) {
        updateItem(
            id: $id
            title: $title
            description: $description
            price: $price
        ) {
            id
            title
            description
            price
        }
    }
`;

export const ITEM_QUERY = gql`
    query ITEM_QUERY(
    $id: ID!
    ) {
        item(where: {
            id: $id
        }) {
            id
            title
            description
            price
            largeImage
        }
    }
`;

export const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION(
    $id: ID!
    ) {
        deleteItem(
            id: $id
        ) {
            id
        }
    }
`;

export const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
    ) {
        signup(
            email: $email
            name: $name
            password: $password
        ) {
            id
            email
            name
        }
    }
`;

export const SIGNIN_MUTATION = gql`
    mutation SIGNIN_MUTATION(
    $email: String!
    $password: String!
    ) {
        signin(
            email: $email
            password: $password
        ) {
            id
            email
            name
        }
    }
`;

export const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY {
        currentUser {
            id
            email
            name
            permissions
			orders {
				id
			}
            cart {
                id
                quantity
                item {
                    id
                    price
                    image
                    title
                    description
                }
            }
        }
    }
`;

export const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }
`;

export const RESET_PASSWORD_MUTATION = gql`
    mutation RESET_PASSWORD_MUTATION(
    $password: String!
    $confirmPassword: String!
    $resetToken: String!
    ) {
        resetPassword(
            password: $password
            confirmPassword: $confirmPassword
            resetToken: $resetToken
        ) {
            id
            email
            name
        }
    }
`;

export const REQUEST_RESET_MUTATION = gql`
    mutation REQUEST_RESET_MUTATION(
    $email: String!
    ) {
        requestReset(
            email: $email
        ) {
            message
        }
    }
`;

export const ALL_USERS_QUERY = gql`
    query ALL_USERS_QUERY {
        users {
            id
            name
            email
            permissions
        }
    }
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
    mutation UPDATE_PERMISSIONS_MUTATION(
    $permissions: [Permission]
    $userId: ID!
    ) {
        updatePermissions(
            permissions: $permissions
            userId: $userId
        ) {
            id
            name
            email
            permissions
        }
    }
`;

// @client tells Apollo to not go to the GraphQL client or the server but check the client state for cartOpen
export const LOCAL_STATE_QUERY = gql`
    query LOCAL_STATE_QUERY {
        cartOpen @client
    }
`;

export const TOGGLE_CART_MUTATION = gql`
    mutation TOGGLE_CART_MUTATION {
        toggleCart @client
    }
`;

export const ADD_TO_CART_MUTATION = gql`
    mutation ADD_TO_CART_MUTATION(
    $id: ID!
    ) {
        addToCart(
            id: $id
        ) {
            id
            quantity
        }
    }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
    mutation REMOVE_FROM_CART_MUTATION(
    $id: ID!
    ) {
        removeFromCart(
            id: $id
        ) {
            id
        }
    }
`;

export const SEARCH_ITEMS_QUERY = gql`
    query SEARCH_ITEMS_QUERY(
    $searchTerm: String!
    ) {
        items(
            where: {
                OR: [
                    { title_contains: $searchTerm },
                    { description_contains: $searchTerm },
                ],

            }
        ) {
			id
			image
			title
		}
    }
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION(
		$token: String!
	) {
		createOrder(
			token: $token
		) {
			id 
			charge
			total
			items {
				id 
				title
			}
		}
	}
`;

export const ORDER_QUERY = gql`
	query ORDER_QUERY (
    $id: ID!
    ) {
		order(
			id: $id
		) {
			id
			items {
				id 
				title
				description
				price
				image
				quantity
			}
			total
			user {
				id
			}
			charge
			createdAt
		}
	}
`;

export const ORDERS_QUERY = gql`
	query ORDERS_QUERY {
		orders(
			OrderBy: createdAt_DESC
		) {
			id
			total
			createdAt
			items {
				id
				title
				price
				description
				quantity
				image
			}
		}
	}
`;