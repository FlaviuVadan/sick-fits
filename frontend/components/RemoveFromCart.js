import React from 'react';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY, REMOVE_FROM_CART_MUTATION } from "../queries/queries";


const RemoveButton = styled.button`
	font-size: 3rem;
	background: none;
	border: 0;
	
	&:hover {
		color: ${props => props.theme.red};
		cursor: pointer;
	}
`;

class RemoveFromCart extends React.Component {
	static propTypes = {
		id: PropTypes.string.isRequired,
	};

	/**
	 * Called upon a response from the back-end
	 * @param cache - Apollo cache
	 * @param payload - content (item id from resolved mutation)
	 */
	update = (cache, payload) => {
		// read cache
		const data = cache.readQuery({ query: CURRENT_USER_QUERY });
		// remove item from cart
		const cartItemId = payload.data.removeFromCart.id;
		// write back to cache
		data.currentUser.cart = data.currentUser.cart.filter(cartItem => cartItem.id !== cartItemId);
		cache.writeQuery({ query: CURRENT_USER_QUERY, data });
	};

	render() {
		return (
			<Mutation
				mutation={REMOVE_FROM_CART_MUTATION}
				variables={{ id: this.props.id }}
				update={this.update}
				// can safely assume the remove request will go through so we do not want to make users wait for it to
				// complete; optimisticResponse used to return a "mock" response from the back-end - in this case, the
				// id of the item to be deleted (used to delete from cache in update upon a successful response)
				optimisticResponse={{
					__typename: 'Mutation', // return type
					removeFromCart: {
						__typename: 'CartItem',
						id: this.props.id,
					}
				}}>
				{(removeFromCart, { loading }) => {
					return (
						<RemoveButton title="Remove Item" disabled={loading}
									  onClick={() => {
										  removeFromCart({}).catch(err => alert(err.message));
									  }}>
							&times;
						</RemoveButton>
					);
				}}
			</Mutation>
		);
	}
}

export default RemoveFromCart;