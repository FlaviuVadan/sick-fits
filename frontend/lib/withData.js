// a high order component that exposes the Apollo client via a prop
// want server side rendering and next-with-apollo helps with that
import withApollo from 'next-with-apollo';

// package by Apollo containing all the std things one may use
// can extend what the Apollo client does and contains a lot of pre-configed things
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';

import { LOCAL_STATE_QUERY } from "../queries/queries";


// header important for authentication
function createClient({ headers }) {
	return new ApolloClient({
		uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
		// include credentials on every request, includes cookies
		request: operation => {
			operation.setContext({
				fetchOptions: {
					credentials: 'include',
				},
				headers,
			});
		},
		// local data
		clientState: {
			resolvers: {
				Mutation: {
					// toggleCart will have one variable garbage collected
					toggleCart(_, variables, { cache }) {
						// read cartOpen value from cache
						const { cartOpen } = cache.readQuery({
							query: LOCAL_STATE_QUERY,
						});
						// write cart state to opposite
						const data = {
							data: {
								cartOpen: !cartOpen,
							},
						};
						cache.writeData(data);
						return data;
					},
				},
			},
			defaults: {
				cartOpen: true
			}
		}
	});
}

export default withApollo(createClient);
