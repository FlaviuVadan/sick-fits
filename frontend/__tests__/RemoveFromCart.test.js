import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import RemoveFromCart from '../components/RemoveFromCart';
import { REMOVE_FROM_CART_MUTATION, CURRENT_USER_QUERY } from "../queries/queries";
import { fakeUser, fakeCartItem } from "../lib/testUtils";

global.alert = jest.fn();

const mocks = [
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: {
				currentUser: {
					...fakeUser(),
					cart: [fakeCartItem({ id: 'abc123' })],
				},
			},
		},
	},
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: {
				currentUser: {
					...fakeUser(),
					cart: [],
				},
			},
		},
	},
	{
		request: {
			query: REMOVE_FROM_CART_MUTATION,
			variables: {
				id: 'abc123',
			},
		},
		result: {
			data: {
				removeFromCart: {
					__typename: 'CartItem',
					id: 'abc123',
				},
			},
		},
	},
];

describe('<RemoveFromtCart/>', () => {
	it('renders', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<RemoveFromCart id="abc123"/>
			</MockedProvider>
		);
		expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
	});

	it('removes item from cart', async () => {
		let apolloClient;
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<ApolloConsumer>
					{client => {
						apolloClient = client;
						return <RemoveFromCart id="abc123"/>;
					}}
				</ApolloConsumer>
			</MockedProvider>
		);
		const response = await apolloClient.query({ query: CURRENT_USER_QUERY });
		expect(response.data.currentUser.cart).toHaveLength(1);
		expect(response.data.currentUser.cart[0].item.price).toBe(5000);

		wrapper.find('button').simulate('click');
		await wait();

		const response2 = await apolloClient.query({ query: CURRENT_USER_QUERY });
		expect(response2.data.currentUser.cart).toHaveLength(0);
	});
});