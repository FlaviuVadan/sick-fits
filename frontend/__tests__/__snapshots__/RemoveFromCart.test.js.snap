// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromtCart/> renders 1`] = `
<button
  className="RemoveFromCart__RemoveButton-emvtd6-0 cjhqJe"
  disabled={false}
  onClick={[Function]}
  title="Remove Item"
>
  Ã—
</button>
`;
